### PR TITLE
parts: add charm plugin (CRAFT-89)

### PR DESCRIPTION
**Pull request by cmatsuoka**
Start integration with craft-parts by adding the charm plugin. This
plugin handles "charm" parts in `charmcraft.yaml`.
    
Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

### CODE REVIEW

File: charmcraft/parts.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+                work_dir=work_dir,
+                cache_dir=cache_dir,
+                venv_dir=venv_dir,
+            )
+            self._lcm.refresh_packages_list()
+        except PartsError as err:
+            raise CommandError(err)
+
+    @property
+    def prime_dir(self):
```
Comment (cjp256) [inline charmcraft/parts.py:195]:
```suggestion
    def prime_dir(self) -> pathlib.Path:
```

File: charmcraft/parts.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+        try:
+            self._lcm = LifecycleManager(
+                {"parts": all_parts},
+                application_name="charmcraft",
+                work_dir=work_dir,
+                cache_dir=cache_dir,
+                venv_dir=venv_dir,
+            )
+            self._lcm.refresh_packages_list()
```
Comment (cjp256) [inline charmcraft/parts.py:190]: Might be an optimization we can do in the future, if using a provided environment we will have a refresh apt cache.

File: tests/test_parts.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+from charmcraft import parts
+
+
+class TestPartValidation:
+    """Part data validation scenarios."""
+
+    def test_part_validation_happy(self):
+        data = {
+            "plugin": "make",
```
Comment (cjp256) [inline tests/test_parts.py:28]: `make` feels a bit odd as it's not owned by charmcraft. perhaps its going to be updated to charm plugin? :)

File: charmcraft/parts.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    if not plugin_name:
+        raise ValueError("'plugin' not defined")
+
+    plugin_class = plugins.get_plugin_class(plugin_name)
+
+    # validate plugin properties
+    plugin_class.properties_class.unmarshal(spec)
+
+    # validate common part properties
+    plugins.strip_plugin_properties(spec, plugin_name=plugin_name)
```
Comment (cjp256) [inline charmcraft/parts.py:165]: I'd maybe make the copy just prior to this to be obvious what's going to nuke it, or comment why the copy is necessary above.  Or perhaps change strip_plugin_properties() to return a new dict instead. No action required, just a thought.

File: tests/test_parts.py (hunk 3 above)
Comment (cmatsuoka) [reply tests/test_parts.py:28]: Changed to test using the charm plugin.

**Review by cjp256 — APPROVED** [review]

**Comment by cmatsuoka** [comment]
Will propose a new version that executes a charm building tool from the plugin.

**Pull request closed without merging**